lib/filewatch/tail_base.rb

```diff
@@ -52,183 +30,165 @@ def initialize(opts={})
-        # but this is a new watched_file
-        # and we can't tell if its contents are the same
+      sdb_key = watched_file.storage_key
+      # NOTE: during initializing the discovered files and sincedb are converted
+      #   but only for old records
+      #   so a watched file might already be in the sincedb
+      sdb_value = @sincedb.find(watched_file, event)
+
+      if sdb_value && sdb_value.watched_file == watched_file
+        yield if block_given? # should open the file
+        return false if !watched_file.file_open?
+        # we have seen this fingerprint before
+        # and it is allocated
+        # its contents are the same
         # as another file we have watched before.
-        last_read_size = @sincedb[sincedb_key]
-        @logger.debug? && @logger.debug("#{path}: sincedb last value #{@sincedb[sincedb_key]}, cur size #{stat.size}")
-        if stat.size > last_read_size
-          # 1) it could really be a new file with lots of new content
-          # 2) it could have old content that was read plus new that is not
-          @logger.debug? && @logger.debug("#{path}: sincedb: seeking to #{last_read_size}")
-          watched_file.file_seek(last_read_size) # going with 2
-          watched_file.update_bytes_read(last_read_size)
-        elsif stat.size == last_read_size
-          # 1) it could have old content that was read
-          # 2) it could have new content that happens to be the same size
-          @logger.debug? && @logger.debug("#{path}: sincedb: seeking to #{last_read_size}")
-          watched_file.file_seek(last_read_size) # going with 1.
-          watched_file.update_bytes_read(last_read_size)
-        else
-          # it seems to be a new file with less content
-          @logger.debug? && @logger.debug("#{path}: last value size is greater than current value, starting over")
-          @sincedb[sincedb_key] = 0
-          watched_file.update_bytes_read(0) if watched_file.bytes_read != 0
+        last_read_size = sdb_value.position
+        @logger.debug? && @logger.debug("_add_to_sincedb: #{path}: #{event}, in sincedb, last value #{last_read_size}, cur size #{stat.size}")
+        case event
+        when :create, :create_initial
+          # a file with the same fingerprint as another has now been allocated to
+          # this sdb_value and it is being processed now.
+          # but we don't want to reread the data
+          @logger.debug? && @logger.debug("_add_to_sincedb: #{path}: #{event}, in sincedb, seeking to #{last_read_size}")
+          watched_file.file_seek(last_read_size)
+          sdb_value.upd_position(last_read_size) # ?? hmmmm: should be the same
+        when :grow
+          # it has old content that was read and more now since the converter matched
+          @logger.debug? && @logger.debug("_add_to_sincedb: #{path}: grow, in sincedb, seeking to #{last_read_size}")
+          watched_file.file_seek(last_read_size)
+          sdb_value.upd_position(last_read_size)
```

Comment:
I understand that `:grow` isn't conceptually the same as **create** and **create_initial**, but it uses the same code?